Improve src dir detection in merlin

rgrinberg opened the pull request:
Currently, the .merlin generation only considers the root directory as
the location of all source dirs. This assumption does not work with
`(include_subdirs ..)`. Instead, we fetch all the modules for local
libraries and then get their directories.

rgrinberg committed:
Improve src dir detection in merlin

Currently, the .merlin generation only considers the root directory as
the location of all source dirs. This assumption does not work with
`(include_subdirs ..)`. Instead, we fetch all the modules for local
libraries and then get their directories.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

aalekseyev commented:
- Do you think you understand the difference between `Lib_info.best_src_dir` and `Lib_info.src_dir`? Why do we use `best_src_dir` in one case and `src_dir` in another?

- Could you add a test case for this, please?

- I think it would be nice to move the new code into a function (e.g. `src_dirs`) somewhere in `Lib` or `Lib_info`, unless there's something specific to merlin in this logic.

rgrinberg commented:
> Do you think you understand the difference between Lib_info.best_src_dir and Lib_info.src_dir? Why do we use best_src_dir in one case and src_dir in another?

Indeed, I should clarify this. best_src_dir is only relevant for external libraries. For local libraries, there's always one choice for the source directory: the directory where the library is defined in the workspace.

> Could you add a test case for this, please?

Yes, I tried doing this but got fed up with how our test suite is organized: 

> I think it would be nice to move the new code into a function (e.g. src_dirs) somewhere in Lib or Lib_info, unless there's something specific to merlin in this logic.

Will do once I write a test confirming the bug is fixed. This entire area of the code needs to be a lot cleaner.

aalekseyev commented:
I'm a bit fuzzy on what the source dir of an external library even means: do we even install sources for external libraries? Do we flatten them if they used `include_subdirs` originally?

rgrinberg commented:
> I'm a bit fuzzy on what the source dir of an external library even means: do we even install sources for external libraries? Do we flatten them if they used include_subdirs originally?

Correct. We install the sources in a flattened directory structure. I guess this only works because module names must be currently unique in a library.

ygrek commented:
I was just going to report this issue, and while searching github found  https://github.com/ocaml/dune/pull/1083, jfyi

aalekseyev commented:
@ygrek, oh, thanks! We should figure out why the existing code doesn't work, then, before adding more.

rgrinberg commented:
It’s a fix for a different problem. We treat source dirs in a single library correctly but we don’t treat source dirs of dependencies correctly. The old code always chose one source dir per dependency . That is clearly wrong
On Jan 2, 2020, 10:26 PM +0700, Arseniy Alekseyev <notifications@github.com>, wrote:
> @ygrek, oh, thanks! We should figure out why the existing code doesn't work, then, before adding more.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

aalekseyev commented:
Oh, I get it now! Thank you.

rgrinberg committed:
Add .merlin generation test

The test is for .merlin files of libraries (and executables) whose
dependencies have more than one source dir.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
@aalekseyev I would refactor this code into Lib_info, but I fear that it would take quite a bit of effort. There's quite a few instances where the source directory is used and it would take a while to verify that I've done everything correctly. Also, I'd really like this fix for 2.1.

Can we postpone the refactor for 2.1.x or 2.2?

aalekseyev commented:
OK. I still moved the function to top-level and added comments linking the two things I was confused about.

aalekseyev commented:
Here's something else I don't understand about the current implementation: is `best_src_dir` useful in the case when the library is local, or could you always write `source_dirs` instead of `Path.Set.union srcs source_dirs`?

rgrinberg commented:
> Here's something else I don't understand about the current implementation: is best_src_dir useful in the case when the library is local, or could you always write source_dirs instead of Path.Set.union srcs source_dirs?

On second glance, I believe it can safely dropped. I was just being overly cautious in my initial fix.

aalekseyev commented:
Ah, that's good. In that case I think we can simplify a bit, and since this is the only call site of `best_src_dir`, we can simplify that thing too (no longer need to think about the local case at all).

rgrinberg commented:
> Ah, that's good. In that case I think we can simplify a bit, and since this is the only call site of best_src_dir, we can simplify that thing too (no longer need to think about the local case at all).

I did this simplification.
